tests/unit/dataframer/dataframer.py

```diff
@@ -0,0 +1,238 @@
+        value_normalized = value['data']
+        value_source = 'valueSampledData'
+    elif 'valueTime' in resource_dict:
+        value_normalized = resource_dict['valueTime']
+        value_source = 'valueTime'
+    elif 'valueDateTime' in resource_dict:
+        value_normalized = resource_dict['valueDateTime']
+        value_source = 'valueDateTime'
+    elif 'valuePeriod' in resource_dict:
+        value = resource_dict['valuePeriod']
+        value_normalized = f"{value['start']} to {value['end']}"
+        value_source = 'valuePeriod'
+    elif 'valueUrl' in resource_dict:
+        value_normalized = resource_dict['valueUrl']
+        value_source = 'valueUrl'
+    elif 'valueDate' in resource_dict:
+        value_normalized = resource_dict['valueDate']
+        value_source = 'valueDate'
+    elif 'valueCount' in resource_dict:
+        value_normalized = resource_dict['valueCount']['value']
+        value_source = 'valueCount'
+    else:
+        value_normalized, value_source = None, None
+        # for debugging...
+        # raise ValueError(f"value[x] not found in {resource_dict}")
+
+    return value_normalized, value_source
+
+
+def normalize_coding(resource_dict: dict) -> list[tuple[str, str]]:
+    def extract_coding(coding_list):
+        # return a concatenated string
+        # or alternatively return an array
+        return [coding.get('display', coding.get('code', '')) for coding in coding_list]
+
+    def find_codings_in_dict(d: dict, parent_key: str = '') -> list[tuple[str, str]]:  # TODO - parent_key not used?
+        codings = []
+        for key, value in d.items():
+            if isinstance(value, list):
+                for item in value:
+                    if isinstance(item, dict):
+                        # Check if the dict contains a 'coding' list
+                        if 'coding' in item and isinstance(item['coding'], list):
+                            coding_string = extract_coding(item['coding'])
+                            codings.append((coding_string, key))
+                        if 'code' in item:
+                            coding_string = item.get('display', item.get('code'))
+                            codings.append((coding_string, key))
+
+                        # Recursively search in the dict
+                        codings.extend(find_codings_in_dict(item, key))
+            elif isinstance(value, dict):
+                # Check if the dict contains a 'coding' list
+                if 'coding' in value and isinstance(value['coding'], list):
+                    coding_string = extract_coding(value['coding'])
+                    codings.append((coding_string, key))
+
+                # Recursively search in the dict
+                codings.extend(find_codings_in_dict(value, key))
+        return codings
+
+    return find_codings_in_dict(resource_dict)
+
+
+def is_number(s):
+    """ Returns True if string is a number. """
+    try:
+        complex(s)
```

Comment:
Had a problem in the data where this check was considering vector/imaginary numbers as numbers. Strings like 4i, and 4j were getting classified as numbers and messing with the typing in elastic/guppy

Reply:
@matthewpeterkort will this work?


```def is_number(s) -> bool:
    """ Returns True if string is a number. """
    methods = [int, float]
    for method in methods:
        # if a number was passed
        if isinstance(s, method):
            return True
        # otherwise it is a string, see if it can be converted to a number
        try:
            method(s)
            return True
        except ValueError:
            pass

    return False


assert is_number('123')
assert is_number('123.456')
assert is_number('0')
assert is_number('0.0')
assert is_number('+0.123')
assert is_number('-123.0')
assert not is_number('- 123')
assert not is_number('+ 123')
assert not is_number('123a')
assert not is_number('a123')
assert not is_number('abc')
assert not is_number('4j')
assert not is_number('4i')

assert is_number(123)
assert is_number(123.456)
```